src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -187,6 +171,22 @@ internal sealed class VisualStudioProject
             _compilationOptions = compilationOptions;
             _filePath = filePath;
             _parseOptions = parseOptions;
+
+            var fileExtensionToWatch = language switch { LanguageNames.CSharp => ".cs", LanguageNames.VisualBasic => ".vb", _ => null };
+
+            if (filePath != null && fileExtensionToWatch != null)
+            {
+                // Since we have a project directory, we'll just watch all the files under that path; that'll avoid extra overhead of
+                // having to add explicit file watches everywhere.
+                var projectDirectoryToWatch = new FileChangeWatcher.WatchedDirectory(Path.GetDirectoryName(filePath), fileExtensionToWatch);
+                _documentFileChangeContext = _workspace.FileChangeWatcher.CreateContext(projectDirectoryToWatch);
+            }
+            else
+            {
+                _documentFileChangeContext = workspace.FileChangeWatcher.CreateContext();
+            }
+
+            _documentFileChangeContext.FileChanged += DocumentFileChangeContext_FileChanged;
```

Comment:
This move fixes it because the document collections need to be instatiated before we react to file changed events.

Reply:
yes

Reply:
I'm not sure if I'm more:

1. ...impressed that you got that unlucky for that to happen.
2. ...terrified we didn't see this sooner, since this code's been around for a few years now.

Reply:
@jasonmalinowski Crazy part is I saw this several times. I was working through a different issue, but after seeing this maybe 5 or 6 times during debugging of the other issue, I went back after the original issue was resolved and fixed this one.